### ISSUE TITLE
Sampling weights in case-control sampling

### ISSUE THREAD

**Issue by lisu-stats** [CAM-Roche/RandomisedTrialsEmulation#36]
Shaun and I have been discussing whether we should apply sampling weights for outcome models fitted with the case-control sampled data. Basically, each control will have a sampling weight equal to (number of potential controls/ number of sampled controls) at each follow-up time for each trial.  @RoonakR @gravesti

**Comment by gravesti** [comment]
@lisu-stats @RoonakR 
Interesting idea. You'll have to explain more about how this works on Monday. I don't think it will be too hard to implement in `case_util()`. I think we can add a new column to the `controlselect` at row 188.

What weight do the cases get? 1?

https://github.com/CAM-Roche/RandomisedTrialsEmulation/blob/e3100a4958b6a52f4e27e37a4171388cd3e47357/R/sampling.R#L173-L192

**Comment by lisu-stats** [comment]
Each case has a weight of one, each selected control has a weight = (**number of potential controls**/ **number of sampled controls**).  @gravesti 

So some lines similar to below (if set n_control=5):
`casedatajk$samplew<-rep(1,ncase)    
      controlselect$samplew<-rep(ncontrol/(ncase*5), ncase*5) `

**Comment by gravesti** [comment]
@lisu-stats How will this work with the other weights, eg in the per protocol analysis?

**Comment by lisu-stats** [comment]
I think we need to multiply them in the per-protocol analyses. I will discuss this with Shaun next Thursday. @gravesti

**Comment by lisu-stats** [comment]
@gravesti @RoonakR  I discussed the case-control sampling issue with Shaun after the Christmas break.  We'd like to propose another case-control algorithm when the baseline hazard as a function of _trial m and follow-up t_ is of interest. 

Basically, in this new algorithm, we also subsample controls for the trials and follow-up visits **where no event of interest occurred**. We sample all controls based on a user-specified probability and a user-specified minimum number of controls to be selected. After the control selection, we then create case-control sampling weights: Each case has a weight of one, each selected control has a weight = (number of potential controls/ number of sampled controls). 

For ITT analyses, we then multiply these case-control sampling weights with any censoring weights.
For per-protocol and as-treated analyses, we take the product of the case-control sampling analyses, the treatment switching weights and censoring weights. 

In this way, we should be able to estimate the baseline hazard function of _trial m and follow-up t_. Also, we shall set 'splines' for fitting these two functions as the default. I will try to provide some sample code.

**Comment by lisu-stats** [comment]
@gravesti Shaun suggested that we change the case-control sampling scheme such that a fixed proportion of controls are selected instead of fixed numbers. The code will similar to this: 

```
library(dplyr)
case_util <- function(data, p_control=0.01){ 
  ### cases occurred at each period and follow-up visit 
  casedatajk<-data[data$outcome==1, ] 
  ### controls (still under follow-up and events haven't occurred) at each period and follow-up visit 
  controldatajk<-data[data$outcome==0, ] 
  dataall = NULL 
  ncase<-dim(casedatajk)[1]  ## number of cases 
  ncontrol<-dim(controldatajk)[1]  ## number of potential controls 
  if(ncase>0){ 
    controlselect<-slice_sample(controldatajk, prop=p_control)
  
    dataall<-rbind(casedatajk, controlselect) ## append sampled data 
  } 
  return(dataall) 
} 
```

**Comment by gravesti** [comment]
@lisu-stats Thanks for this example code. In what setting would we want to have this `if(ncase>0)` condition? I'm wondering how we should control this, if we need another parameter for `initiatiors()`?

**Comment by lisu-stats** [comment]
@lisu-stats If the users only want to estimate the hazard ratios instead of actual survival probabilities (e.g in competing risk setting), then we can keep the condition` if(ncase>0)`

**Comment by gravesti** [comment]
@lisu-stats Is it worth preserving the existing `n_controls` functionality? or is the proportion approach the better way?

**Comment by lisu-stats** [comment]
@gravesti Is it very common to have very few controls when the case occurs in practice? If yes, there is a certain chance that no controls will be selected using the proportion approach. We could set that at least one control needs to be selected.

**Comment by gravesti** [comment]
@lisu-stats With the CPRD data it's very unlikely, however to be safe we can do `ceiling(total_controls * p_controls)` so we always have at least one.

**Comment by lisu-stats** [comment]
@shaun2022 Welcome, Shaun! Do you think we can set a minimum number of selected controls at 1?

**Comment by lisu-stats** [comment]
@gravesti @shaun2022 Isaac, Shaun and I have discussed and agreed that we should sample controls by proportions **_at each trial and follow-up period._** This is because of the specific pooled logistic models we used for the outcome of interest and the Sandwich estimator for the variance. 

We do case-control sampling only because there are too many controls in the data and we can't cope with them due to computational constraints. As we should have enough controls to be sampled, we can remove the minimum number of controls functionality in the code.

Since we sample control by proportions, for those control selected, we shall weight them by 1/proportion when fitting the outcome logistic model. These weights should be multiplied with other treatment and censoring weights when needed. 

For the CPRD analysis, the task is to set the sampling proportion not too small but within the computational constraint.
